File: 2022/11/04/go/reverse_list.go
/**
 * https://leetcode.com/problems/reverse-linked-list/
 * Author: Andrew Jarombek
 * Date: 11/4/2022
 */

package _go

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var result *ListNode = nil
	current := head

	for current != nil {
		node := current
		current = current.Next
		node.Next = result
		result = node
	}

	return result
}

File: 2023/03/30/cpp/reverseList.cpp
// https://leetcode.com/problems/reverse-linked-list/
// Author: Andrew Jarombek
// Date: 3/30/2023

#include <cassert>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* reverseList(ListNode* head) {
    if (head == nullptr || head->next == nullptr)
        return head;

    auto* node = head->next;
    auto* prev = head;

    while (node != nullptr) {
        auto* newHead = node;
        auto* newNext = node->next;

        newHead->next = head;
        head = newHead;

        prev->next = newNext;
        node = newNext;
    }

    return head;
}

int main() {
    auto* node = new ListNode(1,  new ListNode(2));
    auto* result = reverseList(node);

    assert(result->val == 2);
    assert(result->next->val == 1);
    assert(result->next->next == nullptr);
}